Implement new lint: if_then_some_else_none

magurotuna opened the pull request:
Resolves #6760

changelog: Added a new lint: `if_then_some_else_none`

rust-highfive commented:
r? @Manishearth

(rust-highfive has picked a reviewer for you, use r? to override)

giraffate reviewed (commented):
I think we should use msrv for this lint because the suggested [`bool::then`](https://doc.rust-lang.org/std/primitive.bool.html#method.then) was  introduced in `1.50.0`.
https://github.com/rust-lang/rust-clippy/blob/master/doc/adding_lints.md#specifying-the-lints-minimum-supported-rust-version-msrv

magurotuna commented:
@giraffate thanks for pointing that out, that's absolutely correct! Fixed it now.

Manishearth commented:
r? @giraffate do you want to take a shot at reviewing this?

giraffate commented:
Yes, I'll do!

Manishearth commented:
@bors delegate=giraffate

bors commented:
:v: @giraffate can now approve this pull request
<!-- homu: {"type":"Delegated","delegator":"Manishearth","delegate":"giraffate"} -->

giraffate commented on clippy_lints/src/if_then_some_else_none.rs:
```diff
@@ -0,0 +1,105 @@
+            if let ExprKind::Path(ref els_call_qpath) = els_expr.kind;
+            if utils::match_qpath(els_call_qpath, &utils::paths::OPTION_NONE);
+            then {
+                let cond_snip = utils::snippet(cx, cond.span, "[condition]");
+                let arg_snip = utils::snippet(cx, then_arg.span, "");
```
Can test codes like below be added? Using macro, the suggested code may be broken, so it would be better to use the method like `snippet_with_macro_callsite`.
```rust
    let _ = if matches!(true, true) {
        println!("true!");
        Some(matches!(true, false))
    } else {
        None
    };
```

Additionally, can test codes like below be added? The parenthesis isn't considered, so the suggested codes seem not to work.
```rust
    let x = Some(5);
    let _ = x.and_then(|o| if o < 32 { Some(o) } else { None });
```

giraffate commented on clippy_lints/src/if_then_some_else_none.rs:
```diff
@@ -0,0 +1,105 @@
+                let cond_snip = utils::snippet(cx, cond.span, "[condition]");
+                let arg_snip = utils::snippet(cx, then_arg.span, "");
+                let help = format!(
+                    "consider using `bool::then` like: `{}.then(|| {{ /* snippet */ {} }})`",
```
If `then block` has no statements, it would be better to remove the placeholder like `/* snippet */`.

magurotuna commented:
@giraffate 
Thanks a lot for the review!
I addressed your points:

- use `snippet_with_macro_callsite` instead
- make a condition of if expression parenthesized when it is a unary or binary expression
- if a then block has no statements, don't display `/* snippet */` and curly braces of a closure.

Please take a look again :)

giraffate commented:
@bors r+

It looks good. Thanks!

bors commented:
:pushpin: Commit 11d2af7e9694e8e63000e33f77d4926e936e3822 has been approved by `giraffate`

<!-- @bors r=giraffate 11d2af7e9694e8e63000e33f77d4926e936e3822 -->
<!-- homu: {"type":"Approved","sha":"11d2af7e9694e8e63000e33f77d4926e936e3822","approver":"giraffate"} -->

bors commented:
:hourglass: Testing commit 11d2af7e9694e8e63000e33f77d4926e936e3822 with merge 781de34222a7606932a0485670537c1413616246...
<!-- homu: {"type":"BuildStarted","head_sha":"11d2af7e9694e8e63000e33f77d4926e936e3822","merge_sha":"781de34222a7606932a0485670537c1413616246"} -->

bors commented:
:sunny: Test successful - [checks-action_dev_test](https://github.com/rust-lang/rust-clippy/runs/2102861444), [checks-action_remark_test](https://github.com/rust-lang/rust-clippy/runs/2102857463), [checks-action_test](https://github.com/rust-lang/rust-clippy/runs/2102891878)
Approved by: giraffate
Pushing 781de34222a7606932a0485670537c1413616246 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"giraffate","base_ref":"master","builders":{"checks-action_test":"https://github.com/rust-lang/rust-clippy/runs/2102891878","checks-action_dev_test":"https://github.com/rust-lang/rust-clippy/runs/2102861444","checks-action_remark_test":"https://github.com/rust-lang/rust-clippy/runs/2102857463"},"merge_sha":"781de34222a7606932a0485670537c1413616246"} -->

flip1995 commented on clippy_lints/src/if_then_some_else_none.rs:
````diff
@@ -0,0 +1,115 @@
+    ///     42
+    /// });
+    /// ```
+    pub IF_THEN_SOME_ELSE_NONE,
````
Since the MSRV feature is pretty new an we don't have any automated checks for it yet:

If an MSRV is added to a lint, the lint has also to be added in the config option. I wrote documentation about this in #6977. cc @rust-lang/clippy

flip1995 replied:
I will prepare a fix for this (and possible other lints soon.

magurotuna replied:
> If an MSRV is added to a lint, the lint has also to be added in the config option.

Oh I didn't know that. I'll keep that in mind for the next time I implement a new lint.

> I will prepare a fix for this (and possible other lints soon.

Thanks!

giraffate replied:
Thanks for your follow-up!

flip1995 replied:
> Oh I didn't know that. I'll keep that in mind for the next time I implement a new lint.

No worries, this isn't/wasn't documented anywhere, so that's on me for missing it while reviewing the documentation.
